### DISCUSSION TITLE
Self-Relation Cascade in MongoDB

### DISCUSSION THREAD

**Discussion by bekogeko** [prisma/prisma#17746] [Q&A] [answered]
First of all, I want to make things clear about the special rules for referential actions.
the page I read [Special rules for referential actions in SQL Server and MongoDB](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions/special-rules-for-referential-actions)

I have a Comment model which relates to itself to achieve nested comments. It looks like this:
```prisma
model Comment {
    ...
    id String @id @default(auto()) @map("_id") @db.ObjectId
    content String

    parent   Comment?  @relation("ParentChild", fields: [parentId], references: [id], onUpdate: NoAction, onDelete: NoAction)
    children Comment[] @relation("ParentChild")
    parentId String?   @db.ObjectId
    ...
}
```
I have to define the actions to NoAction because of these special rules that apply to MongoDB.
I'm able to add new comments and nest them. But the problem starts when I try to delete one of them. Let me demonstrate the comment relation with an example:
```
Root Comment
- Child Comment #1
- - Baby Comment
- Child Comment #2
```

If I delete Baby Comment or Child Comment #2 it will delete them successfully.  or if we generalize this situation we can delete the comment with no nested comments.

And after that, I tried to delete the relations with child comments. using the below code.
```js
await prisma.comment.update({
    where: {
        id: comment.id,
    },
    data: {
        children: {
            deleteMany: {},
        },
    },
});
```
Finally, I was able to delete both Child Comment #1 too. But the problem doesn't end here.
I still can't delete Root Comment because in order to do that I need to delete the relations even further down.
I'm thinking about some sort of recursive deletion of relations but I didn't find any documentation about it. I even thought about implementing one myself.

Is there any solution to this special rule? I'm open to any advice or question about this method.

using MongoDB.
Prisma Version: 4.8.1

**Comment by nurul3101** [comment, marked as answer]
Hello @bekogeko 👋 
Thank you for writing a detailed description of the scenario. It's immensely helpful in visualising this use case.

> I still can't delete Root Comment because in order to do that I need to delete the relations even further down.

This would mean that you need to first delete all the children nodes (and their children if there are any) before you can delete the root node, right?

I believe a recursive function would be the ideal way here, it could look something similar to this:

```typescript
async function deleteCommentWithChildren(prisma, node) {
  const children = await prisma.comment.findMany({
    where: {
      parent: {
        id: node.id,
      },
    },
  });

  for (const child of children) {
    await deleteCommentWithChildren(prisma, child);
  }

  await prisma.comment.delete({
    where: {
      id: node.id,
    },
  });
}
```
First we need to find all children comments and then call itself for each child comment to delete the child and its descendants, and then finally delete the root comment.

**Reply by bekogeko** [reply]
Wow, thanks a lot! I also tried it but it ended up getting the same error but recursively 😊. 
Is it performant because of all these loops it looks like there are a lot of queries to call? Can't it be achieved with one complex query using the `where` param?

**Reply by nurul3101** [reply]
I haven't tried to optimise this; it should be possible to reduce the number of queries further, but as you mentioned it would make the original query much complex.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
